messaging/models.py

```diff
@@ -74,16 +83,94 @@ class Meta:
     def __str__(self):
         return f"({self.pk}) {self.subject} ({self.sent_at or 'not sent'})"
 
-    def send(self):
-        self.recipient_count = (
-            Guardian.objects.filter(children__project=self.project).count()
-            if self.recipient_selection == Message.ALL
-            else random.randint(
-                0, 50
-            )  # TODO temporary thingy before actual send implementation
+    def send(self, *, force=False):
+        if self.sent_at and not force:
+            raise AlreadySentError()
+
+        guardians = self.get_recipient_guardians()
+
+        self.sent_at = timezone.now()
+        self.recipient_count = len(guardians)
+        self.save(update_fields=("sent_at", "recipient_count"))
+
+        for guardian in guardians:
+            with switch_language(self, guardian.language):
+                if guardian.language in getattr(
+                    settings, "ILMOITIN_TRANSLATED_FROM_EMAIL", {}
+                ):
+                    from_email = settings.ILMOITIN_TRANSLATED_FROM_EMAIL[
+                        guardian.language
+                    ]
+                else:
+                    from_email = settings.DEFAULT_FROM_EMAIL
+
+                send_mail(
+                    self.subject, self.body_text, guardian.email, from_email=from_email,
+                )
+
+    def get_recipient_guardians(self):
+        guardians = Guardian.objects.filter(children__project=self.project)
+
+        if self.recipient_selection == Message.ALL:
+            return guardians.distinct()
+
+        now = timezone.now()
+        upcoming_events = Event.objects.filter(
+            project=self.project, occurrences__time__gte=now
+        ).published()
+        if self.event_id:
+            upcoming_events = upcoming_events.filter(pk=self.event_id)
+
+        children = Child.objects.filter(project=self.project, guardians__isnull=False)
+        enrolments = Enrolment.objects.filter(child__in=children)
+        subscriptions = FreeSpotNotificationSubscription.objects.filter(
+            child__in=children
         )
-        self.sent_at = now()
-        self.save(update_fields=("recipient_count", "sent_at"))
+
+        if self.occurrences.exists():
+            occurrences = self.occurrences.all()
+            enrolments = enrolments.filter(occurrence__in=occurrences)
+            subscriptions = subscriptions.filter(occurrence__in=occurrences)
+        elif self.event_id:
+            occurrences = self.event.occurrences.all()
+            enrolments = enrolments.filter(occurrence__in=occurrences)
```

Comment:
I don't really like how the two lines are repeated there, but because it is possible there are no occurrences, another option would need to be something like

        if self.occurrences.exists():
            occurrences = self.occurrences.all()
        elif self.event_id:
            occurrences = self.event.occurrences.all()
        else:
            occurrences = None
        if occurrences:
            enrolments = enrolments.filter(occurrence__in=occurrences)
            subscriptions = subscriptions.filter(occurrence__in=occurrences)

Which is even more lines 😄

Reply:
ah ok, my bad, didn't know that there is a case `occurrences` is `None` 😥